### PR TITLE
Removed a non-obvious side-effect of assigning Context.template.

### PR DESCRIPTION
**Pull request by aaugustin**
Explicit is better than implicit.

### CODE REVIEW

File: django/template/base.py [hunk 1]
```diff
@@ -204,19 +204,15 @@ def _render(self, context):
...
-        if toplevel_render:
-            context.template = self
         context.render_context.push()
         try:
-            return self._render(context)
+            if context.template is None:
+                with context.bind_template(self):
+                    return self._render(context)
+            else:
+                return self._render(context)
```
Comment (aaugustin) [inline django/template/base.py]: Ideally I would like to call `context.bind_template(self)` unconditionally.

That may become possible if we make the change described in https://github.com/django/django/pull/4161#issuecomment-75136193

File: django/template/context.py [hunk 2]
```diff
@@ -210,28 +221,26 @@ def __init__(self, request, dict_=None, processors=None,
...
-            else:
-                # Set context processors for this engine.
-                processors = (template.engine.template_context_processors +
-                              self._processors)
-                updates = {}
-                for processor in processors:
-                    updates.update(processor(self.request))
-                self.dicts[self._processors_index] = updates
+    @contextmanager
+    def bind_template(self, template):
```
Comment (carljm) [inline django/template/context.py]: You have this defined twice?

Comment (carljm) [inline django/template/context.py]: Er, never mind. Of course you do - failed to see the large chunk of collapsed context in between.

File: django/template/base.py (hunk 1 above)
Comment (prestontimmons) [inline django/template/base.py]: +1. I like this approach.

**Pull request merged**